File: solutions/dart/run-length-encoding/1/lib/run_length_encoding.dart
class RunLengthEncoder {
  String encode(String input) {
    if (input == '') {
      return '';
    }
    String result = "";
    int count = 1;
    String prevChar = input[0];
    for (int i = 1; i < input.length; i++) {
      String char = input[i];
      if (char == prevChar) {
        count++;
      } else {
        if (count == 1) {
          result += prevChar;
        } else {
          result += count.toString();
          result += prevChar;
        }

        count = 1;
        prevChar = char;
      }
    }
    if (count == 1) {
      result += prevChar;
    } else {
      result += count.toString();
      result += prevChar;
    }
    return result;
  }

  String decode(String input) {
    if (input == '') {
      return '';
    }

    String res = '';
    String digits = '';
    if (is_digit(input[0])) {
      digits += input[0];
    } else {
      res += input[0];
    }

    for (int i = 1; i < input.length; i++) {
      if (is_digit(input[i])) {
        digits += input[i];
      } else {
        if (digits == "") {
          res += input[i];
        } else {
          res += input[i] * int.parse(digits);
        }
        digits = '';
      }
    }
    return res;
  }
}

bool is_digit(String c) {
  return c.codeUnitAt(0) ^ 0x30 <= 9;
}

File: solutions/python/run-length-encoding/1/run_length_encoding.py
def decode(string):
    if string=='':
        return ''

    res = ""
    digits = ""
    if string[0].isdigit():
        digits += string[0]
    else:
        res+= string[0]
    for i in range(1, len(string)):
        if string[i].isdigit():
            digits += string[i]
        else:
            if digits == "":
                res+= string[i]
            else:
                n = int(digits)
                res+=string[i]*n
            digits = ""
    
    return res


def encode(string):
    if string=='':
        return ''

    cur_char=string[0]
    cur_count=1

    res=''

    for char in list(string)[1::]:
        if char==cur_char:
            cur_count+=1
        else:
            if cur_count==1:
                res+=cur_char
            else:
                tmp=f'{cur_count}{cur_char}'
                res+=tmp

            cur_char=char
            cur_count=1

    if cur_count==1:
        res+=cur_char
    else:
        tmp=f'{cur_count}{cur_char}'
        res+=tmp

    return res
